qdax/core/emitters/mees_emitter.py

```diff
@@ -0,0 +1,613 @@
+            ),
+            (random_key),
+        )
+
+        return samples, random_key
+
+    @partial(
+        jax.jit,
+        static_argnames=("self",),
+    )
+    def _sample_explore(
+        self,
+        generation_count: int,
+        novelty_archive: Descriptor,
+        repertoire: MapElitesRepertoire,
+        random_key: RNGKey,
+    ) -> Tuple[Genotype, RNGKey]:
+        """Sample uniformly from the 5 most novel genotypes.
+
+        Args:
+            emitter_state: current emitter state
+            repertoire: the current genotypes repertoire
+            random_key: a jax PRNG random key
+
+        Returns:
+            samples: a genotype sampled in the repertoire
+            random_key: an updated jax PRNG random key
+        """
+
+        # Compute the novelty of all indivs in the archive
+        novelties = self._novelty(
+            generation_count + 1,
+            repertoire.descriptors,
+            novelty_archive,
+        )
+        novelties = jnp.where(repertoire.fitnesses > -jnp.inf, novelties, -jnp.inf)
+
+        # Sample uniformaly for the 5 most novel cells
+        max_novelties, _ = jax.lax.top_k(novelties, 5)
+        min_novelty = jnp.nanmin(
+            jnp.where(max_novelties > -jnp.inf, max_novelties, jnp.inf)
+        )
+        repertoire_empty = novelties < min_novelty
+        p = (1.0 - repertoire_empty) / jnp.sum(1.0 - repertoire_empty)
+        random_key, subkey = jax.random.split(random_key)
+        samples = jax.tree_map(
+            lambda x: jax.random.choice(subkey, x, shape=(1,), p=p),
+            repertoire.genotypes,
+        )
+
+        return samples, random_key
+
+    @partial(
+        jax.jit,
+        static_argnames=("self",),
+    )
+    def state_update(
+        self,
+        emitter_state: MEESEmitterState,
+        repertoire: MapElitesRepertoire,
+        genotypes: Genotype,
+        fitnesses: Fitness,
+        descriptors: Descriptor,
+        extra_scores: ExtraScores,
+    ) -> MEESEmitterState:
+        """Generate the gradient offspring for the next emitter call. Also
+        update the novelty archive and generation count from current call.
+
+        Args:
+            emitter_state: current emitter state
+            repertoire: the current genotypes repertoire
+            genotypes: the genotypes of the batch of emitted offspring.
+            fitnesses: the fitnesses of the batch of emitted offspring.
+            descriptors: the descriptors of the emitted offspring.
+            extra_scores: a dictionary with other values outputted by the
+                scoring function.
+
+        Returns:
+            The modified emitter state.
+        """
+
+        assert (
+            jax.tree_util.tree_leaves(genotypes)[0].shape[0] == 1
+        ), "ERROR MAP-Elites-ES generates 1 individual per generation (batch_size=1)."
+
+        ############################
+        # Updating novelty archive #
```

Comment:
I think it would be nicer to have the mechanism hidden in a proper class that implements them. So in this function, we only have to call the method of the class.

Btw, is this way of putting comments a common one?

Reply:
I implemented a NoveltyArchive class at the top of the file following your suggestion.
Concerning the comments, I use them as a way to structure the code when it can be divided in a few main steps, I have no idea if it is standardised. I saw you removed them anyway.